pkg/ignore/dockerignore.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ignore
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+const (
+
+	// dockerIgnoreFile defines the Name of the scanner file
+	dockerIgnoreFile = ".dockerignore"
+	// invertPrefix is used to prefix the Patterns to symbolize inversions
+	invertPrefix = "!"
+)
+
+var (
+	_ FileIgnorer = &defaultIgnorer{}
+	// root directory to start the scanner
+	directory string
+	// patterns holds all the possible Ignorable patterns
+	ignorePatterns []fileIgnorePattern
```

Comment:
These should be fields on the `dockerIgnorer`, global state keeping is usually a "code smell".

Reply:
I was thinking about it , as I hate member variables, as they lead to pandora box

Reply:
just thinking you want to return the defaultIgnorer in all error cases as well ??